include/command_graph_generator.h

```diff
@@ -23,48 +22,52 @@ constexpr size_t max_num_nodes = 256;
  */
 class write_command_state {
-	constexpr static int64_t stale_bit = 1ll << 63;
-	constexpr static int64_t replicated_bit = 1ll << 62;
-	static_assert(sizeof(stale_bit) == sizeof(command_id));
+	friend struct std::hash<celerity::detail::write_command_state>;
+
+	static_assert(alignof(command) > 0b11); // so we have 2 spare bits to encode the masks below
+	constexpr static uintptr_t stale_bit = 0b01;
+	constexpr static uintptr_t replicated_bit = 0b10;
 
   public:
 	constexpr write_command_state() = default;
 
-	/* explicit(false) */ constexpr write_command_state(command_id cid) : m_cid(cid) {}
+	explicit(false) write_command_state(command* const cmd) : m_bits(reinterpret_cast<uintptr_t>(cmd)) {}
 
-	constexpr write_command_state(command_id cid, bool is_replicated) : m_cid(cid) {
-		if(is_replicated) { m_cid |= replicated_bit; }
+	write_command_state(command* const cmd, bool is_replicated) : m_bits(reinterpret_cast<uintptr_t>(cmd)) {
+		if(is_replicated) { m_bits |= replicated_bit; }
 	}
 
-	bool is_fresh() const { return (m_cid & stale_bit) == 0u; }
+	command* get_command() const { return reinterpret_cast<command*>(m_bits & ~stale_bit & ~replicated_bit); }
 
-	bool is_replicated() const { return (m_cid & replicated_bit) != 0u; }
+	bool is_fresh() const { return (m_bits & stale_bit) == 0u; }
 
-	void mark_as_stale() { m_cid |= stale_bit; }
+	bool is_replicated() const { return (m_bits & replicated_bit) != 0u; }
 
-	operator command_id() const { return m_cid & ~stale_bit & ~replicated_bit; }
+	void mark_as_stale() { m_bits |= stale_bit; }
 
-	friend bool operator==(const write_command_state& lhs, const write_command_state& rhs) { return lhs.m_cid == rhs.m_cid; }
+	operator command*() const { return get_command(); }
 
-	friend bool operator!=(const write_command_state& lhs, const write_command_state& rhs) { return !(lhs == rhs); }
+	friend bool operator==(const write_command_state& lhs, const write_command_state& rhs) { return lhs.m_bits == rhs.m_bits; }
+	friend bool operator==(const write_command_state& lhs, const std::nullptr_t /* rhs */) { return lhs.get_command() == nullptr; }
+	friend bool operator==(const std::nullptr_t /* lhs */, const write_command_state& rhs) { return rhs.get_command() == nullptr; }
 
   private:
-	command_id m_cid = 0;
+	uintptr_t m_bits = 0;
 };
 
 class command_graph_generator {
 	friend struct command_graph_generator_testspy;
 
-	inline static const write_command_state no_command = write_command_state(static_cast<command_id>(-1));
+	inline static const write_command_state no_command = {};
 
 	struct buffer_state {
-		buffer_state(region_map<write_command_state> lw, region_map<std::bitset<max_num_nodes>> rr)
-		    : local_last_writer(std::move(lw)), replicated_regions(std::move(rr)), pending_reduction(std::nullopt) {}
+		explicit buffer_state(const range<3>& range, const write_command_state initial_wcs, const node_bitset& replicated_on_nodes)
```

Comment:
Is there a specific reason to make this 3-parameter constructor `explicit`?

Reply:
I just think explicit as the default is good practice.